pkg/cloud/gcp/actuators/util/gcp_tags_labels.go

```diff
@@ -119,3 +127,97 @@ func GetLabelsList(userLabelsAllowed bool, client controllerclient.Client, machi
 
 	return labels, nil
 }
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) []machinev1.ResourceManagerTag {
+	if platformStatus == nil || platformStatus.GCP == nil || platformStatus.GCP.ResourceTags == nil {
+		return nil
+	}
+
+	tags := make([]machinev1.ResourceManagerTag, len(platformStatus.GCP.ResourceTags))
+	for i, tag := range platformStatus.GCP.ResourceTags {
+		tags[i] = machinev1.ResourceManagerTag{
+			ParentID: tag.ParentID,
+			Key:      tag.Key,
+			Value:    tag.Value,
+		}
+	}
+
+	return tags
+}
+
+// getTagValuesNames returns the list of tags in Compute APIs required format, which
+// is a map containing keys of the form Key(`tagKeys/{tag_key_id}`) and values in the
+// form (`tagValues/{tag_value_id}`).
+func getTagValuesNames(ctx context.Context, tagService tagservice.TagService, tagList []machinev1.ResourceManagerTag) (map[string]string, error) {
+	tagValueList := make(map[string]string, len(tagList))
+	for _, tag := range tagList {
+		name := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		value, err := tagService.GetNamespacedName(ctx, name)
+		if err != nil {
+			// fetching tag's metadata could fail due to errors like timeout, server
+			// internal errors, permission issues among others. Since tag's key and
+			// value names are required for binding tag to compute resource, will
+			// return error and retry during next reconciliation.
+			return nil, fmt.Errorf("failed to fetch %s tag details: %w", name, err)
```

Comment:
I actually think we didn't want to drop the "this tag isn't found" section of this. If the error is transient, then retrying is good. If the error is, this tag isn't found, there's no point retrying. You can use an `InvalidMachineConfigurationError` to tell the machine controller to stop trying on permanent errors like that, I think that's what we want here

Reply:
suggestion incorporated.